### PR TITLE
Add types to _dynamo/testing.py

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136402

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

File: torch/_dynamo/testing.py [hunk 1]
```diff
@@ -102,7 +103,7 @@ def requires_bwd_pass(out):
     raise NotImplementedError("Don't know how to reduce", type(out))
 
 
-def reduce_to_scalar_loss(out):
+def reduce_to_scalar_loss(out: Any) -> Union[torch.Tensor, float]:
```
Comment (Skylion007) [inline torch/_dynamo/testing.py:106]: Use an override maybe to tell mypy when the output should be a float vs a Tensor.

**Comment by Skylion007** [comment]
If you want to generally help typing in the codebase, you could try to enable `RUF013`.

**Comment by bobrenjc93** [comment]
> If you want to generally help typing in the codebase generally, you could try to enable `RUF013`.

I assume RUF013 is stricter? My first goal is to get rid of as many `# mypy: allow-untyped-defs` as I can while learning parts of the codebase that I'll likely need to touch to make improvements to dynamic shapes. I'm happy to look into RUF013 after that. Let me know if I misunderstood the intent of your comment!

**Comment by Skylion007** [comment]
> > If you want to generally help typing in the codebase generally, you could try to enable `RUF013`.
> 
> I assume RUF013 is stricter? My first goal is to get rid of as many `# mypy: allow-untyped-defs` as I can while learning parts of the codebase that I'll likely need to touch to make improvements to dynamic shapes. I'm happy to look into RUF013 after that. Let me know if I misunderstood the intent of your comment!

I just forces strict optionals and apply autofixes even in untyped code.

File: torch/_dynamo/testing.py [hunk 2]
```diff
@@ -207,24 +235,26 @@ def __call__(self, gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor])
...
+    def __call__(
+        self, gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]
+    ) -> Callable[..., Any]:
         self.graphs.append(gm)
         return gm.forward
 
 
-def strip_comment(code) -> str:
+def strip_comment(code: str) -> str:
     code = str(code)
```
Comment (Skylion007) [inline torch/_dynamo/testing.py:249]: if the typing is correct, this line is unnecessary.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**